---
title: 'Analysis: Expression levels of PD-L1 (CD274) in lower-grade gliomas (LGG -
  TCGA publicly available data)'
author: "Sebastian Kurscheid"
date: "04 November 2016"
output:
  pdf_document:
    number_sections: yes
    toc: yes
    toc_depth: 3
  html_document:
    fontsize: 12
    highlight: default
    number_section: yes
    theme: flatly
    toc: yes
    toc_depth: 3
---

# Introduction
Purpose of the analysis: Investigate differences in PD-L1 (CD274) expression levels betweeen molecular & pathohistological subgroups of lower grade gliomas (LGG) as well as glioblastoma tumors (GBM), using publicly available from The Canger Genome Atlas (TCGA).

*********

# Methods
TCGA RNA-Seq Level 3 (normalised) data for LGG and GBM samples were obtained through the [NCI GDC portal](https://gdc-portal.nci.nih.gov), and was pre-processed to produce single data frames of expression values with sample annotations (sample type, IDH status, etc). based on supplementary tables 2 & 3 from Ceccarelli et al 2016. For the PD-L1 expression analysis, TCGA GBM and LGG samples were selected for which RNA-Seq data and annotation information were available. For the correlation analysis of PD-L1 expression and promoter CGI CpG methylation levels, a smaller subset of samples was selected due to missigness of GBM samples which have both RNA-Seq and Infinium450k data available.

All pairwise comparisons with two groups were performed using Student's T-Test.
Comparisons with more than two group were compares using ANOVA, with pairwise comparisons performed using Tukey's HSD post-hoc analysis. P-values < 0.5 were considered as statistically significant.

*********
# Results

## Data preparation
```{r loading_external_data, cache=TRUE, tidy=TRUE, tidy.opts=list(width.cutoff=50), echo = FALSE, cache.lazy=FALSE}
require(GenomicRanges)
#-- load the prepared sample annotations --------------
load("~/Data/TCGA/GBM/Ceccarelli_et_al._2016/glioma_annotations.rda")
glioma_annotations$Case <- as.character(glioma_annotations$Case)

#-- load TCGA glioma RNA-Seq data from pre-processing step --------------
load("~/Data/Collaborations/LGG_PDL1/tcga_gliomas_rna_seq.rda")
tcga_gliomas_rna_seq <- log2(tcga_gliomas_rna_seq + 1)

#-- load 450k methylation and annotation data from pre-processing step --------------
load("~/Data/References/Annotations/Platforms/gr.annot450k.rda")
load("~/Data/TCGA/GBM/Infinium450k/tcga.gbm.450k.rda")
load("~/Data/Collaborations/LGG_PDL1/tcga.lgg.450k.rda")
tcga_gliomas_450k <- merge(tcga.lgg.450k, tcga.gbm.450k, by.x = "row.names", by.y = "row.names")

# select only samples which have expression and annotation data available
exp_samples <- intersect(rownames(glioma_annotations), colnames(tcga_gliomas_rna_seq))
meth_samples <- intersect(rownames(glioma_annotations), colnames(tcga_gliomas_450k))
exp_meth_samples <- intersect(exp_samples, meth_samples)
tcga_gliomas_rna_seq <- tcga_gliomas_rna_seq[,exp_samples]
tcga_gliomas_450k <- tcga_gliomas_450k[,meth_samples]
```

Number of available samples for LGG and GBM with expression (RNA-Seq) data, respectively

```{r sample_counts_exp, tidy=TRUE, tidy.opts=list(width.cutoff=50)}
# number of samples with expression data
table(glioma_annotations[exp_samples, ]$Study)

```
Number of available samples for LGG and GBM with methylation (450k) data, respectively

```{r sample_counts_meth, tidy=TRUE, tidy.opts=list(width.cutoff=50)}
# number of samples with methylation data
table(glioma_annotations[meth_samples, ]$Study)

```
Number of available samples for LGG and GBM with methylation (450k) and expression (RNA-Seq) data, respectively

```{r sample_counts_meth_exp, tidy=TRUE, tidy.opts=list(width.cutoff=50)}
# number of samples with expression and methylation data
table(glioma_annotations[exp_meth_samples, ]$Study)

```
*********
### Data preparation
```{r pdl1_idh_mut_wt, tidy=TRUE, tidy.opts=list(width.cutoff=50)}
#-- collate PD-L1 expression levels and tumor annotation data
load(""~/Data/Collaborations/LGG_PDL1/plotData.rda"")
plotData <- as.data.frame(glioma_annotations[exp_samples, c("Study", "Grade","IDH.status", "IDH.codel.subtype", "Supervised.DNA.Methylation.Cluster")])
plotData <- cbind(plotData, as.numeric(tcga_gliomas_rna_seq[grep("CD274", rownames(tcga_gliomas_rna_seq)), exp_samples]))
colnames(plotData)[6] <- "PDL1"
levels(plotData$Study) <- c("LGG", "GBM")
plotData$combined <- NA 
plotData[which(plotData$Study == "GBM" & plotData$IDH.codel.subtype == "IDHwt"),]$combined <- "GBM_IDHwt"
plotData[which(plotData$Study == "GBM" & plotData$IDH.codel.subtype == "IDHmut-non-codel"),]$combined <- "GBM_IDHmut"
plotData[which(plotData$Study == "LGG" & plotData$IDH.codel.subtype == "IDHwt"),]$combined <- "LGG_IDHwt"
plotData[which(plotData$Study == "LGG" & plotData$IDH.codel.subtype == "IDHmut-non-codel"),]$combined <- "LGG_IDHmut_nonCoDel"
plotData[which(plotData$Study == "LGG" & plotData$IDH.codel.subtype == "IDHmut-codel"),]$combined <- "LGG_IDHmut_coDel"
plotData$combined <- as.factor(plotData$combined)
plotData$combined2 <- as.character(plotData$combined)
plotData[grep("GBM", plotData$combined2),]$combined2 <- "GBM"
plotData$combined2 <- as.factor(plotData$combined2)
LGGs <- which(plotData$Study == "LGG")
```

## Comparing PD-L1 expression between Grade II and III samples:
### Boxplot
```{r boxplot_pdl1_grade_2_3, tidy=TRUE, tidy.opts=list(width.cutoff=50)}
#-- Boxplot of PD-L1 expression levels
#-- comparing LGG Grade II & III
# The EntrezID for PD-L1/CD274 is 29126
PDL1 <- grep(29126, rownames(tcga_gliomas_rna_seq))
boxplot(plotData[LGGs,]$PDL1 ~ droplevels(plotData[LGGs,]$Grade),
        axes = FALSE,
        frame = FALSE,
        lwd = 2.75,
        ylim = c(0,10))

axis(1, lwd = 2.75,
     at = c(1,2), 
     labels = c(paste("Grade II\n[N = ", table(droplevels(plotData$Grade))[1], "]", sep = ""),
                paste("Grade III\n[N = ", table(droplevels(plotData$Grade))[2], "]", sep = "")),
     padj = 1,
     cex = 3)

axis(2, lwd = 2.75)
mtext(side = 2, "PD-L1 expression [log2(FPKM + 1)]", line = 2.5, cex = 1)

```

### Statistical Test (t-Test)
```{r t_test_Grade, tidy=TRUE, tidy.opts=list(width.cutoff=50)}
#-- T-test to compare the two population means
t.test(plotData[LGGs,]$PDL1 ~ droplevels(plotData[LGGs,]$Grade))
```

### Conclusions
**There is a statistically significant difference in mean PD-L1 expression levels between Grade II and III tumors.**

*********

## Comparing LGG samples stratified by IDH1/2 mutation status


### Boxplot
```{r boxplot_pdl1_idh_mut_wt, tidy=TRUE, tidy.opts=list(width.cutoff=50)}
#-- PD-L1 expression levels
#-- comparing IDH mut to wt, irrespective of grade --------------
boxplot(plotData[which(plotData$Study == "LGG"),]$PDL1 ~ plotData[which(plotData$Study == "LGG"),]$IDH.status,
        axes = FALSE,
        frame = FALSE,
        lwd = 2.75,
        ylim = c(0,10))

axis(1, lwd = 2.75,
     at = c(1,2), 
     labels = c(paste("IDH mut\n[N = ", table(plotData[which(plotData$Study == "LGG"),]$IDH.status)[1], "]", sep = ""),
                paste("IDH wt\n[N = ", table(plotData[which(plotData$Study == "LGG"),]$IDH.status)[2], "]", sep = "")),
     padj = 1,
     cex = 3)

axis(2, lwd = 2.75)

mtext(side = 2, "PD-L1 expression [log2(FPKM + 1)]", line = 2.5, cex = 1)
```

### Statistical Test (t-Test)
```{r t_test_IDH, tidy=TRUE, tidy.opts=list(width.cutoff=50)}
#-- T-test to compare the two population means
t.test(plotData[which(plotData$Study == "LGG"),]$PDL1 ~ plotData[which(plotData$Study == "LGG"),]$IDH.status)

```
### Conclusions
**There is a statistically significant difference in PD-L1 expression when comparing IDH1/2^mut^ to IDH^wt^ samples, with higher expression levels found in IDH^wt^ samples.**

*********
## Comparing LGG samples stratified by IDH and 1p19q status to GBM
### Boxplot

```{r boxplot_lgg_idh_mut_wt_1p19q_gbm, tidy=TRUE, tidy.opts=list(width.cutoff=50)}
# ToDo : significance indication with horizontal bars
#-- PD-L1 expression levels, 
#-- comparing IDHmut/1p19qcodel to IDmut/1p19qn to IDHwt, including GBM
plotData$combined2 <- factor(as.character(plotData$combined2), levels = levels(plotData$combined2)[ c(2, 3, 4, 1)])
x <- boxplot(plotData$PDL1 ~ plotData$combined2,
        axes = FALSE,
        frame = FALSE,
        lwd = 2.75,
        ylim = c(0,10))

axis(1, lwd = 2.75,
     at = c(1,2,3,4), 
     labels = c(paste("IDH mut,\n1p19q codel\n[N = ", table(plotData$combined2)[1],"]", sep = ""),
                paste("IDH mut,\n1p19q non-codel\n[N = ", table(plotData$combined2)[2],"]", sep = ""),
                paste("IDH wt\n[N = ", table(plotData$combined2)[3], "]", sep = ""),
                paste("GBM\n[N = ", table(plotData$combined2)[4], "]", sep = "")),
     padj = 1)
axis(2, lwd = 2.75)
mtext(side = 2, "PD-L1 expression [log2(FPKM + 1)]", line = 2.5, cex = 1)
```

### Statistical Test (ANOVA)

```{r anova_idh_1p19q_gbm, tidy=TRUE, tidy.opts=list(width.cutoff=50)}
#-- performing ANOVA to check if differences between groups are significant
aov2 <- aov(formula = PDL1 ~ combined2, data = plotData)
# post-hoc analysis shows that they are
TukeyHSD(aov2)
bartlett.test(plotData$PDL1 ~ plotData$combined2)
# but Bartlett test for homogeneity of variances shows that the assumption is violated
# so the ANOVA has to be taken with a grain of salt
# [possibly due to uneven sized groups? or because values are derived from count data?]
```
### Conclusions

**These results show that we observe statistically significant difference in PD-L1 expression levels between the three LGG sub-groups (IDH^mut^ & 1p19q co-del, IDH^mut^ & 1p19q normal, IDH^wt^), while there is no difference between LGG IDH^wt^ and GBM samples.***

*********
## Correlation between PD-L1 promoter CGI CpG methylation and PD-L1 differences

### Collating PD-L1 expression and methylation data
```{r PDL1_methylation_expression_data, tidy=TRUE, tidy.opts=list(width.cutoff=50)}
# GBM data
# get CD274 probes
i1 <- intersect(exp_meth_samples, glioma_annotations[glioma_annotations$Study == "Glioblastoma multiforme",]$Case)
GBM.PDL1.methylation_expression <- as.data.frame(cbind(IDH_status = as.character(glioma_annotations[i1, "IDH.status"]),
                                                       t(tcga.gbm.450k[names(gr.annot450k[grep("CD274", gr.annot450k$UCSC_RefGene_Name)]), i1]),
                                                       CD274 = as.numeric(tcga_gliomas_rna_seq[grep("CD274", rownames(tcga_gliomas_rna_seq)), i1])))
for (i in colnames(GBM.PDL1.methylation_expression[grep("IDH", colnames(GBM.PDL1.methylation_expression), invert = T)])){
  GBM.PDL1.methylation_expression[,i] <- as.numeric(as.character(GBM.PDL1.methylation_expression[,i]))
}
GBM.PDL1.methylation_expression$IDH_status <- as.factor(GBM.PDL1.methylation_expression$IDH_status)
GBM.PDL1.methylation_expression$IDH_status <- relevel(GBM.PDL1.methylation_expression$IDH_status, ref = "WT")

# LGG data
i2 <- intersect(exp_meth_samples, glioma_annotations[glioma_annotations$Study == "Brain Lower Grade Glioma",]$Case)
LGG.PDL1.methylation_expression <- as.data.frame(cbind(IDH_status = as.character(glioma_annotations[i2,]$IDH.status),
                                                       grade = as.character(glioma_annotations[i2,]$Grade),
                                                       histology = as.character(glioma_annotations[i2,]$Histology),
                                                       chr1p19q_status = as.character(glioma_annotations[i2,]$X1p.19q.codeletion),
                                                       t(tcga.lgg.450k[names(gr.annot450k[grep("CD274", gr.annot450k$UCSC_RefGene_Name)]), i2]),
                                                       CD274 = as.numeric(tcga_gliomas_rna_seq[grep("CD274", rownames(tcga_gliomas_rna_seq)), i2])))
for (i in colnames(LGG.PDL1.methylation_expression[grep("cg|CD", colnames(LGG.PDL1.methylation_expression), invert = F)])){
  LGG.PDL1.methylation_expression[,i] <- as.numeric(as.character(LGG.PDL1.methylation_expression[,i]))
}
LGG.PDL1.methylation_expression$IDH_status <- relevel(LGG.PDL1.methylation_expression$IDH_status, ref = "WT")
lgg.idh_mut.codel <- which(LGG.PDL1.methylation_expression$IDH_status == "Mutant" & LGG.PDL1.methylation_expression$chr1p19q_status == "codel")
lgg.idh_mut.non_codel <- which(LGG.PDL1.methylation_expression$IDH_status == "Mutant" & LGG.PDL1.methylation_expression$chr1p19q_status == "non-codel")
lgg.idh_wt <- which(LGG.PDL1.methylation_expression$IDH_status == "WT")
LGG.PDL1.methylation_expression$molecular_subtype <- NA
LGG.PDL1.methylation_expression[lgg.idh_mut.codel,]$molecular_subtype <- "IDHmut_1p19qcodel"
LGG.PDL1.methylation_expression[lgg.idh_mut.non_codel,]$molecular_subtype <- "IDHmut_1p19non-codel"
LGG.PDL1.methylation_expression[lgg.idh_wt,]$molecular_subtype <- "IDHwt"
LGG.PDL1.methylation_expression$molecular_subtype <- as.factor(LGG.PDL1.methylation_expression$molecular_subtype)
LGG.PDL1.methylation_expression$molecular_subtype <- relevel(LGG.PDL1.methylation_expression$molecular_subtype, ref = "IDHwt")

table(LGG.PDL1.methylation_expression$IDH_status, 
      LGG.PDL1.methylation_expression$chr1p19q_status,
      useNA = "always")
```

### XY Plots of PD-L1 expression and DNA methylation (top 2 CpGs) for LGG
```{r LGG_PDL1_methylation_expression_correlation, tidy=TRUE, tidy.opts=list(width.cutoff=50)}
# preparing data
LGGData <- LGG.PDL1.methylation_expression[, grep("cg", colnames(LGG.PDL1.methylation_expression))]
LGGData <- as.matrix(LGGData)
class(LGGData) <- "numeric"

methExpCorLGG <- sapply(colnames(LGG.PDL1.methylation_expression)[grep("cg", colnames(LGG.PDL1.methylation_expression))], function(x){
  IDHwt <- cor(as.numeric(LGG.PDL1.methylation_expression[, x]), 
           log2(as.numeric(LGG.PDL1.methylation_expression[, "CD274"])+1), 
           method = "spearman")
})

#par(mfrow = c(2,1))
methExpXYPlotLGG <- sapply(names(sort(methExpCorLGG)[c(1,2)]), function(x){
  c1 <- cor(as.numeric(LGG.PDL1.methylation_expression[, x]), 
            log2(as.numeric(LGG.PDL1.methylation_expression[, "CD274"])+1), 
            method = "spearman")
  c1 <- round(c1, 2)
  p1 <-plot(as.numeric(LGG.PDL1.methylation_expression[, x]), 
            log2(as.numeric(LGG.PDL1.methylation_expression[, "CD274"])+1),
            xlab = paste("Probe ID ", x, " [beta value]", sep = ""),
            ylab = "PD-L1 [log2(FPKM+1)]",
            xlim = c(0,1),
            ylim = c(0,10),
            main = paste("TCGA LGG\nCorrelation ", c1, " [Spearman]", sep = ""),
            pch = c(16, 17)[as.numeric(glioma_annotations[rownames(LGGData),]$IDH.status)],
            col = c("red", "blue")[as.numeric(glioma_annotations[rownames(LGGData),]$IDH.status)])
  legend("topright", legend = c("IDHmut", "WT"), col = c("red", "blue"), pch = c(16,17))
  abline(lm(log2(as.numeric(LGG.PDL1.methylation_expression[, "CD274"])+1) ~ as.numeric(LGG.PDL1.methylation_expression[, x])), lwd = 2)
  lines(lowess(log2(as.numeric(LGG.PDL1.methylation_expression[, "CD274"])+1) ~ as.numeric(LGG.PDL1.methylation_expression[, x])), col = "green", lwd = 2)
})

```

### XY Plots of PD-L1 expression and DNA methylation (top 2 CpGs) for GBM
```{r GBM_PDL1_methylation_expression_correlation, tidy=TRUE, tidy.opts=list(width.cutoff=50)}
GBMData <- GBM.PDL1.methylation_expression[, grep("cg", colnames(GBM.PDL1.methylation_expression))]
GBMData <- as.matrix(GBMData)
class(GBMData) <- "numeric"


# calculating correlations for GBM data
methExpCorGBM <- sapply(colnames(GBM.PDL1.methylation_expression)[grep("cg", colnames(GBM.PDL1.methylation_expression))], function(x){
  IDHwt <- cor(as.numeric(GBM.PDL1.methylation_expression[, x]), 
               log2(as.numeric(GBM.PDL1.methylation_expression[, "CD274"])+1), 
               method = "spearman")
})


# plot the data for the top 2 CpGs from LGG data
cpgs <- names(sort(methExpCorLGG)[c(1,2)])
#par(mfrow = c(2,1))
methExpXYPlotLGG <- sapply(cpgs, function(x){
  c1 <- cor(as.numeric(GBM.PDL1.methylation_expression[, x]), 
            log2(as.numeric(GBM.PDL1.methylation_expression[, "CD274"])+1), 
            method = "spearman")
  c1 <- round(c1, 2)
  p1 <-plot(as.numeric(GBM.PDL1.methylation_expression[, x]), 
            log2(as.numeric(GBM.PDL1.methylation_expression[, "CD274"])+1),
            xlab = paste("Probe ID ", x, " [beta value]", sep = ""),
            ylab = "PD-L1 [log2(FPKM+1)]",
            xlim = c(0,1),
            ylim = c(0,10),
            main = paste("TCGA GBM\nCorrelation ", c1, " [Spearman]", sep = ""),
            pch = c(16, 17)[as.numeric(glioma_annotations[rownames(GBMData),]$IDH.status)],
            col = c("red", "blue")[as.numeric(glioma_annotations[rownames(GBMData),]$IDH.status)])
  legend("topright", legend = c("IDHmut", "WT"), col = c("red", "blue"), pch = c(16,17))
  abline(lm(log2(as.numeric(GBM.PDL1.methylation_expression[, "CD274"])+1) ~ as.numeric(GBM.PDL1.methylation_expression[, x])), lwd = 2)
  lines(lowess(log2(as.numeric(GBM.PDL1.methylation_expression[, "CD274"])+1) ~ as.numeric(GBM.PDL1.methylation_expression[, x])), col = "green", lwd = 2)
})


```