---
title: 'Analysis: Expression levels of PD-L1 (CD274) in lower-grade gliomas (LGG -
  TCGA publicly available data)'
author: "Sebastian Kurscheid"
date: "04 November 2016"
output:
  html_document:
    fontsize: 12
    highlight: default
    number_section: yes
    theme: flatly
    toc: yes
    toc_depth: 3
  pdf_document:
    number_sections: yes
    toc: yes
    toc_depth: 3
---

# Introduction
Purpose of the analysis: Investigate differences in PD-L1 (CD274) expression levels betweeen molecular & pathohistological subgroups of lower grade gliomas (LGG) as well as glioblastoma tumors (GBM), using publicly available from The Canger Genome Atlas (TCGA).

TCGA RNA-Seq Level 3 (normalised) data for LGG and GBM samples were obtained through the [NCI GDC portal](https://gdc-portal.nci.nih.gov), and was pre-processed to produce single data frames of expression values with sample annotations (sample type, IDH status, etc). based on supplementary tables 2 & 3 from Ceccarelli et al 2016.

*********

# Results

## Data preparation
```{r loading_external_data, cache=TRUE, tidy=TRUE, tidy.opts=list(width.cutoff=50)}
#-- load the prepared sample annotations --------------
load("~/Data/TCGA/GBM/Ceccarelli et al. 2016/glioma_annotations.rda")
glioma_annotations$Case <- as.character(glioma_annotations$Case)

#-- load TCGA glioma RNA-Seq data from pre-processing step --------------
load("~/Data/Collaborations/LGG_PDL1/tcga_gliomas_rna_seq.rda")
tcga_gliomas_rna_seq <- log2(tcga_gliomas_rna_seq + 1)

# select only samples which have expression and annotation data available
i1 <- intersect(rownames(glioma_annotations), colnames(tcga_gliomas_rna_seq))
glioma_annotations <- glioma_annotations[i1,]
tcga_gliomas_rna_seq <- tcga_gliomas_rna_seq[,i1]
```

Number of available samples for LGG and GBM, respectively

```{r sample_counts, tidy=TRUE, tidy.opts=list(width.cutoff=50)}
table(glioma_annotations$Study)

```

*********
### Data preparation
```{r pdl1_idh_mut_wt, tidy=TRUE, tidy.opts=list(width.cutoff=50)}
#-- collate PD-L1 expression levels and tumor annotation data
plotData <- as.data.frame(glioma_annotations[i1, c("Study", "Grade","IDH.status", "IDH.codel.subtype", "Supervised.DNA.Methylation.Cluster")])
plotData <- cbind(plotData, as.numeric(tcga_gliomas_rna_seq[grep("CD274", rownames(tcga_gliomas_rna_seq)), i1]))
colnames(plotData)[6] <- "PDL1"
levels(plotData$Study) <- c("LGG", "GBM")
plotData$combined <- NA 
plotData[which(plotData$Study == "GBM" & plotData$IDH.codel.subtype == "IDHwt"),]$combined <- "GBM_IDHwt"
plotData[which(plotData$Study == "GBM" & plotData$IDH.codel.subtype == "IDHmut-non-codel"),]$combined <- "GBM_IDHmut"
plotData[which(plotData$Study == "LGG" & plotData$IDH.codel.subtype == "IDHwt"),]$combined <- "LGG_IDHwt"
plotData[which(plotData$Study == "LGG" & plotData$IDH.codel.subtype == "IDHmut-non-codel"),]$combined <- "LGG_IDHmut_nonCoDel"
plotData[which(plotData$Study == "LGG" & plotData$IDH.codel.subtype == "IDHmut-codel"),]$combined <- "LGG_IDHmut_coDel"
plotData$combined <- as.factor(plotData$combined)
plotData$combined2 <- as.character(plotData$combined)
plotData[grep("GBM", plotData$combined2),]$combined2 <- "GBM"
plotData$combined2 <- as.factor(plotData$combined2)
LGGs <- which(plotData$Study == "LGG")
```

## Comparing PD-L1 expression between Grade II and III samples:
### Boxplot
```{r boxplot_pdl1_grade_2_3, tidy=TRUE, tidy.opts=list(width.cutoff=50)}
#-- Boxplot of PD-L1 expression levels
#-- comparing LGG Grade II & III
# The EntrezID for PD-L1/CD274 is 29126
PDL1 <- grep(29126, rownames(tcga_gliomas_rna_seq))
boxplot(plotData[LGGs,]$PDL1 ~ droplevels(plotData[LGGs,]$Grade),
        axes = FALSE,
        frame = FALSE,
        lwd = 2.75,
        ylim = c(0,10))

axis(1, lwd = 2.75,
     at = c(1,2), 
     labels = c(paste("Grade II\n[N =", table(droplevels(plotData$Grade))[1], "]", sep = ""),
                paste("Grade III\n[N =", table(droplevels(plotData$Grade))[2], "]", sep = "")),
     padj = 1,
     cex = 3)

axis(2, lwd = 2.75)
mtext(side = 2, "PD-L1 expression [log2(FPKM + 1)]", line = 2.5, cex = 1.25)

```

### Statistical Test (t-Test)
```{r t_test_Grade, tidy=TRUE, tidy.opts=list(width.cutoff=50)}
#-- T-test to compare the two population means
t.test(plotData[LGGs,]$PDL1 ~ droplevels(plotData[LGGs,]$Grade))
```

### Conclusions
**There is a small, but statistically significant difference in mean PD-L1 expression levels between Grade II and III tumors.**

*********

## Comparing LGG samples stratified by IDH1/2 mutation status


### Boxplot
```{r boxplot_pdl1_idh_mut_wt, tidy=TRUE, tidy.opts=list(width.cutoff=50)}
#-- PD-L1 expression levels
#-- comparing IDH mut to wt, irrespective of grade --------------
boxplot(plotData[which(plotData$Study == "LGG"),]$PDL1 ~ plotData[which(plotData$Study == "LGG"),]$IDH.status,
        axes = FALSE,
        frame = FALSE,
        lwd = 2.75,
        ylim = c(0,10))

axis(1, lwd = 2.75,
     at = c(1,2), 
     labels = c(paste("IDH mut\n[N = ", table(plotData[which(plotData$Study == "LGG"),]$IDH.status)[1], "]", sep = ""),
                paste("IDH wt\n[N = ", table(plotData[which(plotData$Study == "LGG"),]$IDH.status)[2], "]", sep = "")),
     padj = 1,
     cex = 3)

axis(2, lwd = 2.75)

mtext(side = 2, "PD-L1 expression [log2(FPKM + 1)]", line = 2.5, cex = 1.25)
```

### Statistical Test (t-Test)
```{r t_test_IDH, tidy=TRUE, tidy.opts=list(width.cutoff=50)}
#-- T-test to compare the two population means
t.test(plotData[which(plotData$Study == "LGG"),]$PDL1 ~ plotData[which(plotData$Study == "LGG"),]$IDH.status)

```
### Conclusions
**There is a statistically significant difference in PD-L1 expression when comparing IDH1/2^mut^ to IDH^wt^ samples, with higher expression levels found in IDH^wt^ samples.**

*********
## Comparing LGG samples stratified by IDH and 1p19q status to GBM
### Boxplot

```{r boxplot_lgg_idh_mut_wt_1p19q_gbm, tidy=TRUE, tidy.opts=list(width.cutoff=50)}
#-- PD-L1 expression levels, 
#-- comparing IDHmut/1p19qcodel to IDmut/1p19qn to IDHwt, including GBM
plotData$combined2 <- factor(as.character(plotData$combined2), levels = levels(plotData$combined2)[ c(2, 3, 4, 1)])
x <- boxplot(plotData$PDL1 ~ plotData$combined2,
        axes = FALSE,
        frame = FALSE,
        lwd = 2.75,
        ylim = c(0,10))

axis(1, lwd = 2.75,
     at = c(1,2,3,4), 
     labels = c(paste("IDH mut,\n1p19q codel\n[N = ", table(plotData$combined2)[1],"]", sep = ""),
                paste("IDH mut,\n1p19q norm\n[N = ", table(plotData$combined2)[2],"]", sep = ""),
                paste("IDH wt\n[N = ", table(plotData$combined2)[3], "]", sep = ""),
                paste("GBM\n[N = ", table(plotData$combined2)[4], "]", sep = "")),
     padj = 1)
axis(2, lwd = 2.75)
mtext(side = 2, "PD-L1 expression [log2(FPKM + 1)]", line = 2.5, cex = 1.25)
```

### Statistical Test (ANOVA)

```{r anova_idh_1p19q_gbm, tidy=TRUE, tidy.opts=list(width.cutoff=50)}
#-- performing ANOVA to check if differences between groups are significant
aov2 <- aov(formula = PDL1 ~ combined2, data = plotData)
# post-hoc analysis shows that they are
TukeyHSD(aov2)
bartlett.test(plotData$PDL1 ~ plotData$combined2)
# but Bartlett test for homogeneity of variances shows that the assumption is violated
# so the ANOVA has to be taken with a grain of salt
# [possibly due to uneven sized groups? or because values are derived from count data?]
```
### Conclusions

**These results show that we observe statistically significant difference in PD-L1 expression levels between the three LGG sub-groups (IDH^mut^ & 1p19q co-del, IDH^mut^ & 1p19q normal, IDH^wt^), while there is no difference between LGG IDH^wt^ and GBM samples.***

*********