---
title: 'Analysis: Expression levels of PD-L1 (CD274) in lower-grade gliomas (LGG -
  TCGA publicly available data)'
author: "Sebastian Kurscheid"
date: "04 November 2016"
output:
  html_document:
    fontsize: 12
    highlight: default
    number_section: yes
    theme: flatly
    toc: yes
    toc_depth: 3
  pdf_document:
    number_sections: yes
    toc: yes
    toc_depth: 3
---

# Introduction
Purpose of the analysis: Investigate differences in PD-L1 (CD274) expression levels betweeen molecular & pathohistological subgroups of lower grade gliomas (LGG) as well as glioblastoma tumors (GBM), using publicly available from The Canger Genome Atlas (TCGA).

TCGA RNA-Seq Level 3 (normalised) data for LGG and GBM samples were obtained through the NCI GDC portal ([link]https://gdc-portal.nci.nih.gov), and was pre-processed to produce single data frames of expression values with sample annotations (sample type, IDH status, etc). based on supplementary tables 2 & 3 from Ceccarelli et al 2016.
*********
# Results
## Data preparation
```{r loading_external_data, cache=TRUE, tidy=TRUE, tidy.opts=(width.cutoff=50)}
#----------  load the prepared sample annotations --------------
load("~/Data/TCGA/GBM/Ceccarelli et al. 2016/glioma_annotations.rda")
glioma_annotations$Case <- as.character(glioma_annotations$Case)

#---------- load TCGA glioma RNA-Seq data from pre-processing step --------------
load("~/Data/Collaborations/LGG_PDL1/tcga_gliomas_rna_seq.rda")
tcga_gliomas_rna_seq <- log2(tcga_gliomas_rna_seq + 1)

# select only samples which have expression and annotation data available
i1 <- intersect(rownames(glioma_annotations), colnames(tcga_gliomas_rna_seq))
glioma_annotations <- glioma_annotations[i1,]
tcga_gliomas_rna_seq <- tcga_gliomas_rna_seq[,i1]
```

Number of available samples for LGG and GBM, respectively

```{r sample_counts, tidy=TRUE, tidy.opts=(width.cutoff=50)}
table(glioma_annotations$Study)

```

*********

## Comparing PD-L1 expression between Grade II and III samples:
### Boxplot
```{r boxplot_pdl1_grade_2_3, tidy=TRUE, tidy.opts=(width.cutoff=50)}
#-------------- Boxplot of PD-L1 expression levels, comparing LGG Grade II & III --------------
# The EntrezID for PD-L1/CD274 is 29126
PDL1 <- grep(29126, rownames(tcga_gliomas_rna_seq))
gradeII <- glioma_annotations[which(glioma_annotations$Grade == "G2"), "Case"]
gradeIII <- glioma_annotations[which(glioma_annotations$Grade == "G3"), "Case"]
boxplot(unlist(tcga_gliomas_rna_seq[PDL1, gradeII]),
        unlist(tcga_gliomas_rna_seq[PDL1, gradeIII]),
        names = c(paste("Grade II\n[N =", length(gradeII), "]", sep = ""),
                  paste("Grade III\n[N =", length(gradeIII), "]", sep = "")),
        xlab = "PD-L1 expression [log2(FPKM + 1)]")
# T-test to compare the two population means
t.test(tcga_gliomas_rna_seq[PDL1, gradeII], 
       tcga_gliomas_rna_seq[PDL1, gradeIII])
```

### Conclusions
**There is a small, but statistically significant difference in mean PD-L1 expression levels between Grade II and III tumors.**

*********

## Comparing LGG samples stratified by IDH1/2 mutation status

### Data preparation
```{r pdl1_idh_mut_wt, tidy=TRUE, tidy.opts=(width.cutoff=50)}
#-------------- collate PD-L1 expression levels and survival data for patients with known IDH1 status --------------
plotData <- as.data.frame(glioma_annotations[i1, c("Study", "IDH.status", "IDH.codel.subtype", "Supervised.DNA.Methylation.Cluster")])
plotData <- cbind(plotData, as.numeric(tcga_gliomas_rna_seq[grep("CD274", rownames(tcga_gliomas_rna_seq)), i1]))
colnames(plotData)[5] <- "PDL1"
levels(plotData$Study) <- c("LGG", "GBM")
plotData$combined <- NA 
plotData[which(plotData$Study == "GBM" & plotData$IDH.codel.subtype == "IDHwt"),]$combined <- "GBM_IDHwt"
plotData[which(plotData$Study == "GBM" & plotData$IDH.codel.subtype == "IDHmut-non-codel"),]$combined <- "GBM_IDHmut"
plotData[which(plotData$Study == "LGG" & plotData$IDH.codel.subtype == "IDHwt"),]$combined <- "LGG_IDHwt"
plotData[which(plotData$Study == "LGG" & plotData$IDH.codel.subtype == "IDHmut-non-codel"),]$combined <- "LGG_IDHmut_nonCoDel"
plotData[which(plotData$Study == "LGG" & plotData$IDH.codel.subtype == "IDHmut-codel"),]$combined <- "LGG_IDHmut_coDel"
plotData$combined <- as.factor(plotData$combined)
plotData$combined2 <- as.character(plotData$combined)
plotData[grep("GBM", plotData$combined2),]$combined2 <- "GBM"
plotData$combined2 <- as.factor(plotData$combined2)
```

### Boxplot
```{r boxplot_pdl1_idh_mut_wt, tidy=TRUE, tidy.opts=(width.cutoff=50)}
#-------------- PD-L1 expression levels, comparing IDH mut to wt, irrespective of grade --------------
boxplot(plotData[which(plotData$Study == "LGG"),]$PDL1 ~ plotData[which(plotData$Study == "LGG"),]$IDH.status,
        axes = FALSE,
        frame = FALSE,
        lwd = 2.75,
        ylim = c(0,10))

axis(1, lwd = 2.75,
     at = c(1,2), 
     labels = c(paste("IDH mut\n[N = ", table(plotData[which(plotData$Study == "LGG"),]$IDH.status)[1], "]", sep = ""),
                paste("IDH wt\n[N = ", table(plotData[which(plotData$Study == "LGG"),]$IDH.status)[2], "]", sep = "")),
     padj = 1)

axis(2, lwd = 2.75)

mtext(side = 2, "PD-L1 expression [log2(FPKM + 1)]", line = 2.5, cex = 1.25)
```

### Statistical Test (t-Test)
```{r}
# T-test to compare the two population means
t.test(plotData[which(plotData$Study == "LGG"),]$PDL1 ~ plotData[which(plotData$Study == "LGG"),]$IDH.status)

```
### Conclusions
**There is a statistically significant difference in PD-L1 expression when comparing IDH1/2^mut^ to IDH^wt^ samples, with higher expression levels found in IDH^wt^ samples.**

*********

## Comparing LGG CIMP positive to negative samples
### Boxplot

```{r boxplot_cimp_pos_neg, tidy=TRUE, tidy.opts=(width.cutoff=50)}
#-------------- PD-L1 expression levels, comparing CIMP pos to neg, irrespective of grade --------------
CIMPpos <- rownames(plotData[grep("CIMP", plotData$Supervised.DNA.Methylation.Cluster),])
CIMPneg <- rownames(plotData[grep("CIMP", invert = T, plotData$Supervised.DNA.Methylation.Cluster),])
boxplot(unlist(tcga_gliomas_rna_seq[PDL1, CIMPpos]),
        unlist(tcga_gliomas_rna_seq[PDL1, CIMPneg]),
        names = c("CIMP+", "CIMP-"),
        xlab = "PD-L1 expression [log2(FPKM + 1)]")
```

### Statistical Test (t-Test)

```{r}
# T-test to compare the two population means
t.test(unlist(plotData[CIMPpos, "PDL1"]),
        unlist(plotData[CIMPneg, "PDL1"]))

```
### Conclusions

**The previous finding is replicated when looking at CIMP^+/-^ status instead of IDH^mut/wt^ - CIMP^+^ status is causally linked with IDH1/2 mutation and could therefore be used as a proxy. This is important, as CIMP status is available for nearly twice as many samples as IDH mutation status.**

*********

## Comparing LGG samples stratified by IDH and 1p19q status to GBM
### Boxplot

```{r boxplot_lgg_idh_mut_wt_1p19q_gbm, tidy=TRUE, tidy.opts=(width.cutoff=50)}
#-------------- PD-L1 expression levels, comparing IDHmut/1p19qcodel to IDmut/1p19qn to IDHwt, including GBM  --------------
plotData$combined2 <- factor(as.character(plotData$combined2), levels = levels(plotData$combined2)[ c(2, 3, 4, 1)])
boxplot(plotData$PDL1 ~ plotData$combined2,
        axes = FALSE,
        frame = FALSE,
        lwd = 2.75,
        ylim = c(0,10))

axis(1, lwd = 2.75,
     at = c(1,2,3,4), 
     labels = c(paste("IDH mut,\n1p19q codel\n[N = ", table(plotData$combined2)[1],"]", sep = ""),
                paste("IDH mut,\n1p19q norm\n[N = ", table(plotData$combined2)[2],"]", sep = ""),
                paste("IDH wt\n[N = ", table(plotData$combined2)[3], "]", sep = ""),
                paste("GBM\n[N = ", table(plotData$combined2)[4], "]", sep = "")),
     padj = 1)
axis(2, lwd = 2.75)
mtext(side = 2, "PD-L1 expression [log2(FPKM + 1)]", line = 2.5, cex = 1.25)
```

### Statistical Test (ANOVA)

```{r}
# performing ANOVA to check if differences between groups are significant
aov2 <- aov(formula = PDL1 ~ combined2, data = plotData)
# post-hoc analysis shows that they are

TukeyHSD(aov2)
bartlett.test(plotData$PDL1 ~ plotData$combined2)
# but Bartlett test for homogeneity of variances shows that the assumption is violated
# so the ANOVA has to be taken with a grain of salt
# [possibly due to uneven sized groups? or because values are derived from count data?]
```
### Conclusions

**These results show that we observe statistically significant difference in PD-L1 expression levels between the three LGG sub-groups (IDH^mut^ & 1p19q co-del, IDH^mut^ & 1p19q normal, IDH^wt^), while there is no difference between LGG IDH^wt^ and GBM samples.***

*********

## Comparing LGG samples stratified by CIMP and 1p19q status to GBM
### Boxplot

```{r boxplot_lgg_cimp_1p19q_gbm, tidy=TRUE, tidy.opts=(width.cutoff=50)}
#-------------- collate PD-L1 expression levels and survival data for patients with known CIMP status [more CIMP cases then known IDH1/2 status] --------------
CIMPpos1p19qcodel <- which(plotData$Supervised.DNA.Methylation.Cluster %in% c("G-CIMP-high", "G-CIMP-low") & plotData$combined2 == "LGG_IDHmut_coDel")
CIMPpos1p19qnorm <- glioma_annotations[which(glioma_annotations$hCIMP == 1 & glioma_annotations$co1p19q == "n"), "bcr_patient_barcode"]
CIMPpos <- glioma_annotations[which(glioma_annotations$hCIMP == 1), "bcr_patient_barcode"]
CIMPneg <- glioma_annotations[which(glioma_annotations$hCIMP == 0), "bcr_patient_barcode"]

PDL1.CIMPstatusKnownPatients  <- c(unlist(tcga_gliomas_rna_seq[PDL1, CIMPpos1p19qcodel]), 
                                  unlist(tcga_gliomas_rna_seq[PDL1, CIMPpos1p19qnorm]),
                                  unlist(tcga_gliomas_rna_seq[PDL1, CIMPneg]))
PDL1.CIMPstatusKnownPatients <- data.frame("PDL1exp" = PDL1.CIMPstatusKnownPatients, 
                                          "status" = c(rep("CIMPpos1p19qcodel", length(CIMPpos1p19qcodel)), 
                                                       rep("CIMPpos1p19qnorm", length(CIMPpos1p19qnorm)),
                                                       rep("CIMPneg", length(CIMPneg))),
                                          "months_to_death" = glioma_annotations[c(CIMPpos1p19qcodel, CIMPpos1p19qnorm, CIMPneg), "days_to_death"]/ 30)
PDL1.CIMPstatusKnownPatients$death <- 0
PDL1.CIMPstatusKnownPatients[!is.na(PDL1.CIMPstatusKnownPatients$months_to_death), "death"] <- 1

#-------------- PD-L1 expression levels, comparing CIMPpos/1p19qcodel to CIMPpos/1p19qn to CIMPneg, including GBM  --------------
tab1 <- PDL1.CIMPstatusKnownPatients[, c("PDL1exp", "status")]
tab2 <- data.frame("PDL1exp" = unlist(tcga_gliomas_rna_seq[PDL1,]), "status" = rep("GBM", ncol(tcga_gliomas_rna_seq)))
tab1 <- rbind(tab1, tab2)
tab1$status <- factor(as.character(tab1$status), levels = c("CIMPpos1p19qcodel", "CIMPpos1p19qnorm", "CIMPneg", "GBM"))

boxplot(tab1$PDL1exp ~ tab1$status,
        axes = FALSE,
        frame = FALSE,
        lwd = 2.75,
        ylim = c(0,10))

axis(1, lwd = 2.75,
     at = c(1,2,3,4), 
     labels = c(paste("CIMP+,\n1p19q codel\n[N = ", length(CIMPpos1p19qcodel),"]", sep = ""),
               paste("CIMP+,\n1p19q norm\n[N = ", length(CIMPpos1p19qnorm),"]", sep = ""),
               paste("CIMP-\n[N = ", length(CIMPneg), "]", sep = ""),
               paste("GBM\n[N = ", ncol(tcga_gliomas_rna_seq), "]", sep = "")),
     padj = 1)
axis(2, 
     lwd = 2.75)
mtext(side = 2, "PD-L1 expression [log2 FPKM]", line = 2.5, cex = 1.25)

```

### Statistical Test (ANOVA)

```{r}
# performing ANOVA to check if differences between groups are significant
aov2 <- aov(formula = PDL1exp ~ status, data = tab1)
# post-hoc analysis shows that they are

TukeyHSD(aov2)
bartlett.test(tab1$PDL1exp ~ tab1$status)
# but the Bartlett test for homogeneity of variances shows that the assumption is violated
# so the ANOVA has to be taken with a grain of salt
# [possibly due to uneven sized groups? or because values are derived from count data?]
```
### Conclusions

**These results indicate that there is a statistically significant difference when comparing mean PD-L1 expression levels between CIMP^+^ [both 1p19 normal and co-deleted] and CIMP^-^ LGG samples as well as GBM samples. However, there is significant difference between the expression levels in GBMs and CIMP^-^ LGGs. After correcting a coding error in the analysis, I can now confirm that the trend the same trend in PD-L1 expression differences is observed between 1p19q co-deleted and normal subgroups, irrespective of using IDHmut or CIMP status.**

*********

# Appendix
## Data pre-processing
Included for documentation purpose.
```{r pre-processing, echo = TRUE, cache = TRUE, eval = FALSE, tidy=TRUE, tidy.opts=(width.cutoff=50)}
#---------- Expression & DNA methylation data on GDU cluster
path <- "/Volumes/gduserv/Data/TCGA/LGG"

#---------- Import list of LGG RNA-Seq samples --------------
lgg.RNASeq.samples <- read.table(paste(path, "METADATA", "UNC__IlluminaHiSeq_RNASeqV2", "unc.edu_LGG.IlluminaHiSeq_RNASeqV2.1.15.0.sdrf.txt", sep = "/"), header = T, as.is = T, sep = "\t")
lgg.RNASeq.samples$sample_id <- unlist(lapply(strsplit(lgg.RNASeq.samples$Comment..TCGA.Barcode., "-"), function(x) paste(x[1:3], collapse = "-")))

# only retain rows with RSEM gene-level normalized data files & filter out duplicated IDs
lgg.RNASeq.samples <- lgg.RNASeq.samples[grep("RSEM_genes_normalized", lgg.RNASeq.samples$Protocol.REF.4),]
lgg.RNASeq.samples <- lgg.RNASeq.samples[!duplicated(lgg.RNASeq.samples$sample_id),]

# get the unique files
lgg.sampleIDs <-lgg.RNASeq.samples$sample_id
lgg.rna_seq_files <- lgg.RNASeq.samples$Derived.Data.File
names(lgg.rna_seq_files) <- lgg.sampleIDs

#---------- TCGA LGG RNA-Seq data processing --------------
path.rna_seq <- paste(path, "/RNASeqV2/UNC__IlluminaHiSeq_RNASeqV2/Level_3/", sep = "")

# using normalized data
# PD-L1 (CD274) Entrez gene ID: 29126
PDL1 <- grep(29126, rownames(tcga_gliomas_rna_seq))
for (i in 1:length(lgg.rna_seq_files)){
  if (i == 1){
    temp <- read.table(paste(path.rna_seq, lgg.rna_seq_files[i], sep = ""), header = T, as.is = T, sep = "\t")
  } else {
    temp2 <- read.table(paste(path.rna_seq, lgg.rna_seq_files[i], sep = ""), header = T, as.is = T, sep = "\t")
    temp <- cbind(temp, temp2$normalized_count)
  }
}
lgg.rna_seq.norm <- temp
rm(temp)
rownames(lgg.rna_seq.norm) <- lgg.rna_seq.norm$gene_id
lgg.rna_seq.norm <- lgg.rna_seq.norm[,-1]
colnames(lgg.rna_seq.norm) <- names(lgg.rna_seq_files)
# save(lgg.rna_seq.norm, file = "~/Data/Collaborations/LGG_PDL1/lgg.rna_seq.rda")
# we only retain data for patients with clinical data
lgg.rna_seq.norm <- lgg.rna_seq.norm[,which(colnames(lgg.rna_seq.norm) %in% rownames(glioma_annotations))]
# log2 transformation of data, adding offset of 1 to avoid -Inf
tcga_gliomas_rna_seq <- log2(lgg.rna_seq.norm + 1)
# save(tcga_gliomas_rna_seq, file = "~/Data/Collaborations/LGG_PDL1/tcga_gliomas_rna_seq.rda")
# PD-L1 expression levels:
# summary(unlist(tcga_gliomas_rna_seq[grep(29126, rownames(tcga_gliomas_rna_seq)),]))
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 0.000   2.830   3.674   3.660   4.358   8.285
# human PD-L1 (CD274, EntrezID: 29126) is located on chr 9p24

#---------- Import list of GBM RNA-Seq samples --------------
gbmPath <- "/home/skurscheid/Data/TCGA/GBM"
gbm.RNASeq.samples <- read.table(paste(gbmPath, "METADATA", "UNC__IlluminaHiSeq_RNASeqV2", "unc.edu_GBM.IlluminaHiSeq_RNASeqV2.1.4.0.sdrf.txt", sep = "/"), header = T, as.is = T, sep = "\t")
gbm.RNASeq.samples$sample_id <- unlist(lapply(strsplit(gbm.RNASeq.samples$Comment..TCGA.Barcode., "-"), function(x) paste(x[1:3], collapse = "-")))

# only retain rows with RSEM gene-level normalized data files & filter out duplicated IDs
gbm.RNASeq.samples <- gbm.RNASeq.samples[grep("RSEM_genes_normalized", gbm.RNASeq.samples$Protocol.REF.4),]
gbm.RNASeq.samples <- gbm.RNASeq.samples[!duplicated(gbm.RNASeq.samples$sample_id),]

# get the unique files
gbm.sampleIDs <-gbm.RNASeq.samples$sample_id
gbm.rna_seq_files <- gbm.RNASeq.samples$Derived.Data.File
names(gbm.rna_seq_files) <- gbm.sampleIDs

#---------- TCGA LGG RNA-Seq data processing --------------
path.rna_seq <- paste(gbmPath, "/RNASeqV2/UNC__IlluminaHiSeq_RNASeqV2/Level_3/", sep = "")

# using normalized data
# PD-L1 (CD274) Entrez gene ID: 29126
for (i in 1:length(gbm.rna_seq_files)){
  if (i == 1){
    temp <- read.table(paste(path.rna_seq, gbm.rna_seq_files[i], sep = ""), header = T, as.is = T, sep = "\t")
  } else {
    temp2 <- read.table(paste(path.rna_seq, gbm.rna_seq_files[i], sep = ""), header = T, as.is = T, sep = "\t")
    temp <- cbind(temp, temp2$normalized_count)
  }
}
gbm.rna_seq.norm <- temp
rm(temp)
rownames(gbm.rna_seq.norm) <- gbm.rna_seq.norm$gene_id
gbm.rna_seq.norm <- gbm.rna_seq.norm[,-1]
colnames(gbm.rna_seq.norm) <- names(gbm.rna_seq_files)
# save(lgg.rna_seq.norm, file = "~/Data/Collaborations/LGG_PDL1/lgg.rna_seq.rda")
# we only retain data for patients with clinical data
# log2 transformation of data, adding offset of 1 to avoid -Inf
tcga_gliomas_rna_seq <- log2(gbm.rna_seq.norm + 1)
# save(tcga_gliomas_rna_seq, file = "~/Data/Collaborations/LGG_PDL1/tcga_gliomas_rna_seq.rda")
# PD-L1 expression levels:
# summary(unlist(tcga_gliomas_rna_seq[grep(29126, rownames(tcga_gliomas_rna_seq)),]))
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 0.000   2.830   3.674   3.660   4.358   8.285
# human PD-L1 (CD274, EntrezID: 29126) is located on chr 9p24

```
